src/cascadia/TerminalSettingsModel/FileUtils.cpp

```diff
@@ -39,87 +43,57 @@ namespace winrt::Microsoft::Terminal::Settings::Model
-    {
-        wil::unique_hfile file{ CreateFileW(path.c_str(), GENERIC_READ | GENERIC_WRITE, FILE_SHARE_READ | FILE_SHARE_WRITE, nullptr, OPEN_ALWAYS, FILE_ATTRIBUTE_NORMAL, nullptr) };
-        THROW_LAST_ERROR_IF(!file);
-        // just lock the entire file
-        OVERLAPPED sOverlapped;
-        sOverlapped.Offset = 0;
-        sOverlapped.OffsetHigh = 0;
-        // Shared lock
-        THROW_LAST_ERROR_IF(!LockFileEx(file.get(),
-                                        LOCKFILE_EXCLUSIVE_LOCK, // lock exclusive, wait to return until lock is obtained
-                                        0, // reserved, does nothing
-                                        INT_MAX, // lock INT_MAX bytes
-                                        0, // higher-order bytes, if our state file is greater than 2GB I guess this will be a problem
-                                        &sOverlapped));
-        return { std::move(file), sOverlapped };
-    }
-
-    std::string ReadUTF8FileLocked(const locked_hfile& file)
-    {
-        const auto fileSize = GetFileSize(file.get(), nullptr);
-        THROW_LAST_ERROR_IF(fileSize == INVALID_FILE_SIZE);
-
-        // By making our buffer just slightly larger we can detect if
-        // the file size changed and we've failed to read the full file.
-        std::string buffer(static_cast<size_t>(fileSize) + 1, '\0');
-        DWORD bytesRead = 0;
-        THROW_IF_WIN32_BOOL_FALSE(ReadFile(file.get(), buffer.data(), gsl::narrow<DWORD>(buffer.size()), &bytesRead, nullptr));
-
-        // As mentioned before our buffer was allocated oversized.
-        buffer.resize(bytesRead);
-
-        if (til::starts_with(buffer, Utf8Bom))
-        {
-            // Yeah this memmove()s the entire content.
-            // But I don't really want to deal with UTF8 BOMs any more than necessary,
-            // as basically not a single editor writes a BOM for UTF8.
-            buffer.erase(0, Utf8Bom.size());
-        }
-
-        return buffer;
-    }
-
-    void WriteUTF8FileLocked(const locked_hfile& file, const std::string_view& content)
+    // Tries to read a file somewhat atomically without locking it.
+    // Strips the UTF8 BOM if it exists.
+    std::string ReadUTF8File(const std::filesystem::path& path, const bool elevatedOnly)
     {
-        // truncate the file because we want to overwrite it
-        SetFilePointer(file.get(), 0, nullptr, FILE_BEGIN);
-        THROW_IF_WIN32_BOOL_FALSE(SetEndOfFile(file.get()));
-
-        const auto fileSize = gsl::narrow<DWORD>(content.size());
-        DWORD bytesWritten = 0;
-        THROW_IF_WIN32_BOOL_FALSE(WriteFile(file.get(), content.data(), fileSize, &bytesWritten, nullptr));
-
-        if (bytesWritten != fileSize)
+        if (elevatedOnly)
         {
-            THROW_WIN32_MSG(ERROR_WRITE_FAULT, "failed to write whole file");
+            const bool hadExpectedPermissions{ _isOwnedByAdministrators(path) };
+            if (!hadExpectedPermissions)

```

Comment:
Okay, one last thing. We may be introducing a "TOCTOU" (Time of check to time of use) vulnerability.

The user is allowed to delete files in this directory (Medium IL), and in so doing overwrite `elevated-state`.
Therefore: if we pass this check, there's a point at which the user can delete it and replace it with a non-Admin-owned copy here after line 95.

I think the correct thing to do would be to always open the file first thing (including the `DELETE` sharing type: we hold a handle, if somebody deletes it, we will continue to read from and check the original file), THEN check the ownership using `GetSecurityInfo` (not `Named`) and the `HANDLE`, and only then read it from the same handle.